### ISSUE TITLE
🐛 Multiple entries in the list of formatters

### ISSUE THREAD

**Issue by momargoh** [biomejs/biome#1584]
## what happened
1. Install Biome VSCode extension v2.1.1 in a devcontainer and put it in devcontainer extensions to be auto-installed.
2. Open preferences (Cmd+Shift+P), "format document with..." -> Biome appears twice in the list.
3. Rebuild the devcontainer -> Biome appears another two times every time I rebuild
 
## what is expected
Biome VSCode extension should only appear once in the list of formatters and should not increase every time the devcontainer is rebuilt.

**Comment by arendjr** [comment]
Just now I had again 3 formatter entries, and one thing I do notice is that I sometimes I also get a message from VS Code saying it cannot connect to the Biome daemon. It's not often, and when it happened earlier today I checked and there was still only one formatter entry. But just now it happened again, and then there were three. I cannot say it's the cause, but it is something suspicious.

I can see the following extension log inside VS Code:

```
Biome binary found at /home/arendjr/Projects/monofiber/node_modules/@biomejs/cli-linux-x64/biome
Copying binary to temporary folder: file:///home/arendjr/.config/Code/User/workspaceStorage/27f9d18133d60f7c90f1954460f241d6/biomejs.biome/biome
Executing Biome from: /home/arendjr/.config/Code/User/workspaceStorage/27f9d18133d60f7c90f1954460f241d6/biomejs.biome/biome
[cli-stdout] data 24
[cli] exit 0
Connecting to "/tmp/biome-socket-1.3.3" ...
[Info  - 7:25:08 PM] Server initialized with PID: 2770276
Reloading biome executable.
Biome binary found at /home/arendjr/Projects/monofiber/node_modules/@biomejs/cli-linux-x64/biome
[cli-stderr] end
[cli-stderr] finish
[cli-stderr] close
[cli-stdout] end
[cli-stdout] finish
[cli] close 0
[cli-stdout] close
Copying binary to temporary folder: file:///home/arendjr/.config/Code/User/workspaceStorage/27f9d18133d60f7c90f1954460f241d6/biomejs.biome/biome
Executing Biome from: /home/arendjr/.config/Code/User/workspaceStorage/27f9d18133d60f7c90f1954460f241d6/biomejs.biome/biome
[cli-stdout] data 24
[cli] exit 0
Connecting to "/tmp/biome-socket-1.3.3" ...
[Info  - 7:31:06 PM] Server initialized with PID: 2777362
Reloading biome executable.
Reloading biome executable.
Biome binary found at /home/arendjr/Projects/monofiber/node_modules/@biomejs/cli-linux-x64/biome
Biome binary found at /home/arendjr/Projects/monofiber/node_modules/@biomejs/cli-linux-x64/biome
Copying binary to temporary folder: file:///home/arendjr/.config/Code/User/workspaceStorage/27f9d18133d60f7c90f1954460f241d6/biomejs.biome/biome
[cli-stderr] end
[cli-stderr] finish
[cli-stdout] end
[cli-stdout] finish
[cli-stdout] close
[cli] close 0
[cli-stderr] close
Copying binary to temporary folder: file:///home/arendjr/.config/Code/User/workspaceStorage/27f9d18133d60f7c90f1954460f241d6/biomejs.biome/biome
Error copying file: EntryNotFound (FileSystemError): Unable to delete nonexistent file '/home/arendjr/.config/Code/User/workspaceStorage/27f9d18133d60f7c90f1954460f241d6/biomejs.biome/biome'
Executing Biome from: /home/arendjr/Projects/monofiber/node_modules/@biomejs/cli-linux-x64/biome
Executing Biome from: /home/arendjr/.config/Code/User/workspaceStorage/27f9d18133d60f7c90f1954460f241d6/biomejs.biome/biome
[cli-stdout] data 24
[cli] exit 0
[cli-stdout] data 24
[cli-stdout] end
[cli-stdout] finish
[cli-stderr] end
[cli-stderr] finish
[cli] exit 0
Connecting to "/tmp/biome-socket-1.3.3" ...
[cli-stderr] close
[cli] close 0
[cli-stdout] close
[Info  - 7:31:44 PM] Server initialized with PID: 2777766
Reloading biome executable.
Biome binary found at /home/arendjr/Projects/monofiber/node_modules/@biomejs/cli-linux-x64/biome
[cli-stderr] end
[cli-stderr] finish
[cli-stderr] close
[cli-stdout] end
Connecting to "/tmp/biome-socket-1.3.3" ...
[cli-stdout] finish
[Error - 7:31:45 PM] Biome client: couldn't create connection to server.
Error: connect ECONNREFUSED /tmp/biome-socket-1.3.3
	at PipeConnectWrap.afterConnect [as oncomplete] (node:net:1494:16)
Reloading client failed: Error: Could not connect to the Biome server at "/tmp/biome-socket-1.3.3": connect ECONNREFUSED /tmp/biome-socket-1.3.3
[cli] close 0
[cli-stdout] close
Copying binary to temporary folder: file:///home/arendjr/.config/Code/User/workspaceStorage/27f9d18133d60f7c90f1954460f241d6/biomejs.biome/biome
Executing Biome from: /home/arendjr/.config/Code/User/workspaceStorage/27f9d18133d60f7c90f1954460f241d6/biomejs.biome/biome
[cli-stdout] data 24
[cli] exit 0
Reloading biome executable.
Biome binary found at /home/arendjr/Projects/monofiber/node_modules/@biomejs/cli-linux-x64/biome
Copying binary to temporary folder: file:///home/arendjr/.config/Code/User/workspaceStorage/27f9d18133d60f7c90f1954460f241d6/biomejs.biome/biome
Executing Biome from: /home/arendjr/.config/Code/User/workspaceStorage/27f9d18133d60f7c90f1954460f241d6/biomejs.biome/biome
[cli-stdout] data 24
Reloading biome executable.
Biome binary found at /home/arendjr/Projects/monofiber/node_modules/@biomejs/cli-linux-x64/biome
Copying binary to temporary folder: file:///home/arendjr/.config/Code/User/workspaceStorage/27f9d18133d60f7c90f1954460f241d6/biomejs.biome/biome
[cli-stderr] end
[cli-stderr] finish
[cli-stdout] end
[cli-stdout] finish
[cli] exit 0
Connecting to "/tmp/biome-socket-1.3.3" ...
[cli-stdout] close
[cli] close 0
[cli-stderr] close
[Info  - 7:31:45 PM] Server initialized with PID: 2777809
Executing Biome from: /home/arendjr/.config/Code/User/workspaceStorage/27f9d18133d60f7c90f1954460f241d6/biomejs.biome/biome
[cli-stdout] data 24
[cli-stderr] end
[cli-stderr] finish
[cli-stderr] close
[cli-stdout] end
[cli-stdout] finish
[cli] exit 0
Connecting to "/tmp/biome-socket-1.3.3" ...
[cli] close 0
[cli-stdout] close
[Info  - 7:31:45 PM] Server initialized with PID: 2777809
Connecting to "/tmp/biome-socket-1.3.3" ...
[Info  - 7:31:46 PM] Server initialized with PID: 2777809
```

As you can see the server does seem to restart every now and then, though the log doesn't say why unfortunately.

**Comment by nhedger** [comment]
@arendjr the client reloads when a change is detected in a lockfile.

We do this to handle the user updating biome via the cli.

**Comment by arendjr** [comment]
Hmm, so I just tried `touch yarn.lock` a whole bunch of times, and indeed it reloaded many times. Still have only one formatter entry though, so no dice...

**Comment by ematipico** [comment]
I think the issue may be related to where we create the socket.

It's the only change related to the VSCode extension and the LSP that I could think of: https://github.com/biomejs/biome/pull/1238

**Comment by nhedger** [comment]
As suspected, reverting #1104 fixes it for me.

Yet to investigate why exactly that happens.

**Comment by ematipico** [comment]
Maybe both dynamic and static registrations go in conflict?

**Comment by nhedger** [comment]
That would be my guess, too.

**Comment by nhedger** [comment]
Alright, so using the advertised client capabilities, I've conditionally registered to the providers statically. I'll send a PR.

**Comment by ematipico** [comment]
We just published a nightly with a possible fix: https://github.com/biomejs/biome/releases/tag/cli/v1.5.2-nightly.91c7730

It would be great if someone can try it, and let us know.

**Comment by snigdha920** [comment]
> We just published a nightly with a possible fix: https://github.com/biomejs/biome/releases/tag/cli/v1.5.2-nightly.91c7730
> 
> It would be great if someone can try it, and let us know.

Thank you, works for me with: 

```
Name: Biome (nightly)
Id: biomejs.biome-nightly
Description: Biome LSP VS Code Extension
Version: 2023.12.31946
Publisher: biomejs
VS Marketplace Link: https://marketplace.cursorapi.com/items?itemName=biomejs.biome-nightly
```

**Comment by SpasiboKojima** [comment]
> We just published a nightly with a possible fix: https://github.com/biomejs/biome/releases/tag/cli/v1.5.2-nightly.91c7730
> 
> It would be great if someone can try it, and let us know.

I still have it in v.1.8.3, and I think I had it all along this year, is there anything else that can be done?

**Comment by ematipico** [comment]
> > We just published a nightly with a possible fix: [`cli` (release)](https://github.com/biomejs/biome/releases/tag/cli/v1.5.2-nightly.91c7730)
> > It would be great if someone can try it, and let us know.
> 
> I still have it in v.1.8.3, and I think I had it all along this year, is there anything else that can be done?

OP and other people haven't replied in long time, so we assumed everything was fixed. I suggest you to create a new issue with a reproduction, and we can look into it.

**Comment by jay-l-e-e** [comment]
I have the same problem in 1.8.4.nightly build (and earlier versions too). I don't know under what conditions it occurs. This issue caused the code to become corrupted when auto-saving in VScode. Sometimes in an import statement, sometimes in the middle of code, etc.

But when this problem occurs there are always multiple formatter instances. (There appears to be a conflict between several biome extension clients.)

My running environment is Windows 11 Pro (23H2) and my biome settings are:

`{
  "$schema": "https://biomejs.dev/schemas/1.8.3/schema.json",
  "files": {
    "include": [
      "./sources/**/*.ts",
      "./sources/**/*.js",
      "./sources/**/*.graphql",
      "./scripts/**/*.ts",
      "./scripts/**/*.js",
      "./**/*.json"
    ]
  },
  "formatter": {
    "enabled": true,
    "formatWithErrors": false,
    "indentStyle": "space",
    "indentWidth": 2,
    "lineEnding": "lf",
    "lineWidth": 80,
    "attributePosition": "auto"
  },
  "organizeImports": { "enabled": true },
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true,
      "complexity": {
        "noStaticOnlyClass": "off"
      },
      "correctness": {
        "noUnusedVariables": "error",
        "useArrayLiterals": "error"
      },
      "style": {
        "useBlockStatements": "error",
        "useConsistentArrayType": "error",
        "useForOf": "error",
        "useShorthandAssign": "error"
      }
    }
  },
  "overrides": [
    {
      "include": [
        "./sources/common/generations/**/*.ts",
        "./scripts/tests/**/*.ts"
      ],
      "linter": {
        "rules": {
          "complexity": {
            "noBannedTypes": "off"
          },
          "suspicious": {
            "noExplicitAny": "off"
          }
        }
      }
    }
  ],
  "javascript": {
    "formatter": {
      "jsxQuoteStyle": "double",
      "quoteProperties": "asNeeded",
      "trailingCommas": "all",
      "semicolons": "always",
      "arrowParentheses": "always",
      "bracketSpacing": true,
      "bracketSameLine": false,
      "quoteStyle": "single",
      "attributePosition": "auto"
    }
  }
}
`

**Comment by jay-l-e-e** [comment]
> > > We just published a nightly with a possible fix: [`cli` (release)](https://github.com/biomejs/biome/releases/tag/cli/v1.5.2-nightly.91c7730)
> > > It would be great if someone can try it, and let us know.
> > 
> > 
> > I still have it in v.1.8.3, and I think I had it all along this year, is there anything else that can be done?
> 
> OP and other people haven't replied in long time, so we assumed everything was fixed. I suggest you to create a new issue with a reproduction, and we can look into it.

I have the same problem in the 1.8.4 nightly build (and earlier versions too). I don't know under what conditions it occurs. This issue causes the code to become corrupted when auto-saving in VSCode. Sometimes it breaks the import statements, sometimes it breaks the middle of the code, etc.

However, whenever this problem occurs, there are "always" multiple formatter instances running. (It seems like a conflict between multiple biome extension clients.)

My running environment is Windows 11 Pro (23H2) and my biome settings are:

`{
  "$schema": "https://biomejs.dev/schemas/1.8.3/schema.json",
  "files": {
    "include": [
      "./sources/**/*.ts",
      "./sources/**/*.js",
      "./sources/**/*.graphql",
      "./scripts/**/*.ts",
      "./scripts/**/*.js",
      "./**/*.json"
    ]
  },
  "formatter": {
    "enabled": true,
    "formatWithErrors": false,
    "indentStyle": "space",
    "indentWidth": 2,
    "lineEnding": "lf",
    "lineWidth": 80,
    "attributePosition": "auto"
  },
  "organizeImports": { "enabled": true },
  "linter": {
    "enabled": true,
    "rules": {
      "recommended": true,
      "complexity": {
        "noStaticOnlyClass": "off"
      },
      "correctness": {
        "noUnusedVariables": "error",
        "useArrayLiterals": "error"
      },
      "style": {
        "useBlockStatements": "error",
        "useConsistentArrayType": "error",
        "useForOf": "error",
        "useShorthandAssign": "error"
      }
    }
  },
  "overrides": [
    {
      "include": [
        "./sources/common/generations/**/*.ts",
        "./scripts/tests/**/*.ts"
      ],
      "linter": {
        "rules": {
          "complexity": {
            "noBannedTypes": "off"
          },
          "suspicious": {
            "noExplicitAny": "off"
          }
        }
      }
    }
  ],
  "javascript": {
    "formatter": {
      "jsxQuoteStyle": "double",
      "quoteProperties": "asNeeded",
      "trailingCommas": "all",
      "semicolons": "always",
      "arrowParentheses": "always",
      "bracketSpacing": true,
      "bracketSameLine": false,
      "quoteStyle": "single",
      "attributePosition": "auto"
    }
  }
}
`

**Comment by manzaloros** [comment]
I get multiple Biome entries in my VSCode tooltip windows:

<img width="461" alt="Screenshot 2024-09-27 at 2 41 29 PM" src="https://github.com/user-attachments/assets/29bd220d-00fc-404a-8cc2-dc1cc37e3755">

biome json:
```
{
  "$schema": "https://biomejs.dev/schemas/1.9.2/schema.json",
  "organizeImports": {
    "enabled": true
  },
  "files": {
    "ignore": [
      "node_modules",
      "server/protos",
      "build",
      ".yarn/releases",
      "./scripts",
      "assets",
      "public"
    ]
  },
  "linter": {
    "enabled": true,

    "rules": {
      "recommended": true,
      "correctness": {
        "noUnusedVariables": "error",
        "noUnusedImports": "error",
        "noVoidElementsWithChildren": "error",
        "useExhaustiveDependencies": "off",
        "useHookAtTopLevel": "error"
      },
      "suspicious": {
        "noConsoleLog": "error",
        "noExplicitAny": "off"
      },
      "style": {
        "useBlockStatements": "error",
        "noDefaultExport": "error",
        "noRestrictedGlobals": {
          "level": "error",
          "options": {
            "deniedGlobals": ["fetch"]
          }
        }
      },
      "nursery": {
        "noRestrictedImports": {
          "level": "error",
          "options": {
            "paths": {
              "dayjs": "Use the custom dayjs wrapper in @common-libs/utils",
              "moment": "Use the custom dayjs wrapper in @common-libs/utils",
              "moment-timezone": "Use the custom dayjs wrapper in @common-libs/utils"
            }
          }
        },
        "useComponentExportOnlyModules": {
          "level": "error",
          "options": {
            "allowExportNames": [
              "json",
              "loader",
              "headers",
              "meta",
              "links",
              "scripts",
              "action",
              "useHydrated"
            ]
          }
        }
      }
    }
  },
  "formatter": {
    "enabled": true,
    "formatWithErrors": false,
    "ignore": [],
    "attributePosition": "auto",
    "indentStyle": "space",
    "indentWidth": 2,
    "lineEnding": "lf",
    "lineWidth": 80
  },
  "json": {
    "formatter": {
      "trailingCommas": "none"
    }
  },
  "overrides": [
    {
      "include": [
        ... multiple items here...
      ],
      "linter": {
        "rules": {
          "style": {
            "noDefaultExport": "off"
          }
        }
      }
    }
  ]
}

```

**Comment by sairion** [comment]
I am still experiencing this issue. Using Biome 1.8.3 and VS Code extension (v2.3.0) if it helps. It occurs very unexpectedly (which I find it hard to reproduce) and get fixed after completely exit VS Code and open again.

**Comment by LIlGG** [comment]
I have Dev Containers in my VS Code extension, and the problem disappeared after I uninstalled it. But I'm not sure why.

**Comment by LucHermkens** [comment]
This issue is still active for my team. Is there any definitive solution? Disabling the 'Dev Containers' extension solved the problem for me, but not for others I work with, and it's not a viable solution in my opinion. My team has unfortunately switched back to ESLint because of this. I would like this issue to be reopened. I'm using the latest Cursor version 0.46.5 (VSCode 1.96.2).

**Comment by jpinn97** [comment]
I'm using Biome in all my projects now, using the dev container extension too. I have a NX monorepo of which I would have dozens of Biome formatters available that would lock up the window on saving.. Uninstalling the dev container extension fixed that for me, this problem wasn't an issue with my other repos.

**Comment by maxktz** [comment]
I face that every day running Biome 1.9.4

**Comment by ignissak** [comment]
How is still not fixed? :/ still happening in 1.9.4

**Comment by matyaszednicek** [comment]
**Fixed it**: added `"biome.requireConfiguration": true` to settings - like this it will only start one instance in the dir that has `biome.json`

**Comment by WerdoxDev** [comment]
This is still happening in 2.0.0 using the v3 of the extension. Very frustrating. I have a monorepo project with a multiple root vscode workspace. It logs that it started 11 instances and that would be correct but in each project I always see the error list or suggestions twice. Then applying a suggestion runs it twice which for sorting imports, it always breaks the entire imports with random stuff. Basically unusable.

In task manager I see 11 (maybe 12?) instances of biome which is correct since I have 11 projects in the monorepo
![Image](https://github.com/user-attachments/assets/6bcdc7e5-4553-4893-a136-18add6cc28be)

But somehow all suggestions, errors, formatter, is shows twice
![Image](https://github.com/user-attachments/assets/7dbdcace-44ba-4bc3-b7de-7b58897c62a5)

**Comment by ematipico** [comment]
Have you followed the migration guide @WerdoxDev ?

**Comment by qimmer** [comment]
This still an issue for me using 2.0.4 and 2.0.5 with 3.0.3 extension. Biome constantly corrupts my files on save, probably because multiple instances running fight to write the same file. The amount of instances spawned seem to explode over time, sometimes reaching 30+ instances.

![Image](https://github.com/user-attachments/assets/5d8c3d43-9656-479e-b6c6-a396b143fa86)

**Comment by ematipico** [comment]
@qimmer 

Have you found a _consistent_ way to reproduce this? If so, can you tell us how?

**Comment by qimmer** [comment]
It's actually a day-old fresh SolidStart project using SolidJS's own project template with Tailwind, added authjs and a docker-compose.yml and it already started happening here. It happens on another SolidStart project I have as well. Running the newest VSCode and Code OSS on Manjaro Linux, installed the latest 3.0.3 extension from both the Open VSX store on OSS and the official store on VSCode. @biomejs/biome version is 2.0.5 and 2.0.4 has same problem. It usually takes 5-10 minutes before instances start to duplicate. Has been a problem for almost 6 months now.

Pay attention to that number of instances can ALSO go down:
```
2025-06-24 07:42:01.576 [info]  ✅ Biome is ready.
2025-06-24 07:42:16.511 [info]  🔒 Lockfile "/home/kim/Desktop/tylstrup-appen/bun.lock" created.
2025-06-24 07:42:16.511 [info]  🔄 Restarting Biome...
2025-06-24 07:42:16.511 [info]  🔒 Lockfile "/home/kim/Desktop/tylstrup-appen/bun.lock" created.
2025-06-24 07:42:16.511 [info]  🔄 Restarting Biome...
2025-06-24 07:42:17.518 [info]  🔒 Started listening for lockfile changes.
2025-06-24 07:42:17.518 [info]  ⚙️ Started listening for configuration changes.
2025-06-24 07:42:17.519 [info]  🔒 Started listening for lockfile changes.
2025-06-24 07:42:17.519 [info]  ⚙️ Started listening for configuration changes.
2025-06-24 07:42:17.539 [info]  ✅ Biome is ready.
2025-06-24 07:42:17.576 [info]  ✅ Biome is ready.
2025-06-24 07:42:47.876 [info]  🔒 Lockfile "/home/kim/Desktop/tylstrup-appen/bun.lock" created.
2025-06-24 07:42:47.876 [info]  🔄 Restarting Biome...
2025-06-24 07:42:47.876 [info]  🔒 Lockfile "/home/kim/Desktop/tylstrup-appen/bun.lock" created.
2025-06-24 07:42:47.876 [info]  🔄 Restarting Biome...
2025-06-24 07:42:48.881 [info]  🔒 Started listening for lockfile changes.
2025-06-24 07:42:48.881 [info]  ⚙️ Started listening for configuration changes.
2025-06-24 07:42:48.881 [info]  🔒 Started listening for lockfile changes.
2025-06-24 07:42:48.881 [info]  ⚙️ Started listening for configuration changes.
2025-06-24 07:42:48.903 [info]  ✅ Biome is ready.
2025-06-24 07:42:48.911 [info]  ✅ Biome is ready.
2025-06-24 07:48:47.042 [info]  🔒 Started listening for lockfile changes.
2025-06-24 07:48:47.042 [info]  ⚙️ Started listening for configuration changes.
2025-06-24 07:48:47.103 [info]  ✅ Biome is ready.
2025-06-24 07:54:33.522 [info]  🔒 Lockfile "/home/kim/Desktop/tylstrup-appen/bun.lock" created.
2025-06-24 07:54:33.523 [info]  🔄 Restarting Biome...
2025-06-24 07:54:33.523 [info]  🔒 Lockfile "/home/kim/Desktop/tylstrup-appen/bun.lock" created.
2025-06-24 07:54:33.524 [info]  🔄 Restarting Biome...
2025-06-24 07:54:33.524 [info]  🔒 Lockfile "/home/kim/Desktop/tylstrup-appen/bun.lock" created.
2025-06-24 07:54:33.524 [info]  🔄 Restarting Biome...
2025-06-24 07:54:34.524 [info]  🔒 Started listening for lockfile changes.
2025-06-24 07:54:34.524 [info]  ⚙️ Started listening for configuration changes.
2025-06-24 07:54:34.524 [info]  🔒 Started listening for lockfile changes.
2025-06-24 07:54:34.524 [info]  ⚙️ Started listening for configuration changes.
2025-06-24 07:54:34.542 [info]  🔒 Started listening for lockfile changes.
2025-06-24 07:54:34.542 [info]  ⚙️ Started listening for configuration changes.
2025-06-24 07:54:34.600 [info]  ✅ Biome is ready.
2025-06-24 07:54:34.605 [info]  ✅ Biome is ready.
2025-06-24 07:54:34.612 [info]  ✅ Biome is ready.
2025-06-24 08:05:46.123 [info]  🔒 Started listening for lockfile changes.
2025-06-24 08:05:46.123 [info]  ⚙️ Started listening for configuration changes.
2025-06-24 08:05:46.189 [info]  ✅ Biome is ready.
2025-06-24 08:05:58.071 [info]  ⚙️ Configuration changed.
2025-06-24 08:05:58.071 [info]  🔄 Restarting Biome...
2025-06-24 08:05:59.135 [info]  🔒 Started listening for lockfile changes.
2025-06-24 08:05:59.135 [info]  ⚙️ Started listening for configuration changes.
2025-06-24 08:05:59.283 [info]  ✅ Biome is ready.
2025-06-24 08:24:46.296 [info]  🔒 Lockfile "/home/kim/Desktop/tylstrup-appen/bun.lock" created.
2025-06-24 08:24:46.296 [info]  🔄 Restarting Biome...
2025-06-24 08:24:46.297 [info]  🔒 Lockfile "/home/kim/Desktop/tylstrup-appen/bun.lock" created.
2025-06-24 08:24:46.297 [info]  🔄 Restarting Biome...
2025-06-24 08:24:47.301 [info]  🔒 Started listening for lockfile changes.
2025-06-24 08:24:47.301 [info]  ⚙️ Started listening for configuration changes.
2025-06-24 08:24:47.301 [info]  🔒 Started listening for lockfile changes.
2025-06-24 08:24:47.301 [info]  ⚙️ Started listening for configuration changes.
2025-06-24 08:24:47.377 [info]  ✅ Biome is ready.
2025-06-24 08:24:47.389 [info]  ✅ Biome is ready.
```
It also often cause code duplication/corruption like this (notice the excess </DropdownMenu> and succeeding 4 lines):
```typescript
>
      {(user) => (
        <DropdownMenu>
          <DropdownMenuTrigger as={Button} variant="default">
            <Icon iconId="user" />
            {user().name}
          </DropdownMenuTrigger>
          <DropdownMenuContent>
            <DropdownMenuItem>
              <Icon iconId="door-exit" />
              {t.logout}
            </DropdownMenuItem>
          </DropdownMenuContent>
        </DropdownMenu>
      )}
    </Show>
  );
}
        </DropdownMenu>
      )}
    </Show>
  );
}

```
For convenience, here is my settings.json for vscode:
```json
{
  "workbench.iconTheme": "material-icon-theme",
  "workbench.colorTheme": "Dracula Theme",
  "editor.fontSize": 16,
  "editor.fontFamily": "'Fixedsys Excelsior', 'Droid Sans Mono', 'monospace', monospace",
  "editor.tabSize": 2,
  "editor.lineHeight": 1,
  "editor.defaultFormatter": "biomejs.biome",
  "editor.formatOnSave": true,
  "editor.codeActionsOnSave": {
    "source.fixAll.biome": "explicit",
    "source.organizeImports.biome": "explicit",
  },
  "diffEditor.experimental.showMoves": true,
  "diffEditor.experimental.useTrueInlineView": true,
  "workbench.panel.showLabels": false,
  "explorer.confirmDragAndDrop": false,
  "typescript.updateImportsOnFileMove.enabled": "always",
  "javascript.updateImportsOnFileMove.enabled": "always",
  "explorer.confirmDelete": false,
  "git.enableSmartCommit": true,
  "debug.onTaskErrors": "debugAnyway",
  "git.autofetch": true,
  "terminal.integrated.tabs.enabled": false,
  "biome.suggestInstallingGlobally": false,
  "biome.requireConfiguration": true
}
```

**Comment by arendjr** [comment]
It's hard for us to create fix for this since it seems _most_ people don't have the issue and we're unsure what triggers it, but I think at least we should reopen this to acknowledge the issue.

**Comment by qimmer** [comment]
I might have a suspicion if it could be related to multiple CTRL+S's in a row when fixes are applied? But only a hunch.

**Comment by qimmer** [comment]
I have published my repo as public for you guys to try out. Note it might take a couple of modifications and tries before it starts to spawn parallel instances.

[https://github.com/qimmer/tylstrup-appen.git](https://github.com/qimmer/tylstrup-appen.git)

**Comment by ematipico** [comment]
My gut tells me that this issue might occur when users upgrade Biome package version from their editor.

A new connection gets created and the old one isn't shut down. However, Zed and Intellij don't have these issues, as well as other editor extensions out there.

So there's something going on in VSCode too.

@qimmer I've seen your reproduction, but I couldn't find instructions to try and trigger the issue _from the editor_. How can we do that?

**Comment by kimatik** [comment]
@ematipico I have not upgraded anything from within the editor and it happens on a fresh editor restart as well.
I simply open my editor, edit a couple of .tsx files and at somepoint when I save and it should optimize imports on save, it corrupts the file.

**Comment by qimmer** [comment]
@ematipico I have not upgraded anything from within the editor and it happens on a fresh editor restart as well.
I simply open my editor (vscode or oss), open the repo directory, "bun i" to install packages and edit a couple of .tsx files and at somepoint when I save and it should optimize imports on save, it corrupts the file. The Biome extension is installed from the extensions panel and vscode detects the biome version from my package.json.

**Comment by ematipico** [comment]
@nhedger 

I did some triaging. The issue belongs to the VSCode only. I still don't know what the cause is, but if you open the settings file and change `"biome.requireConfiguration": true,` from one value to another, you will notice that a new process named `biome` is spawned. Steps:
- change the value to the setting
- wait for Biome to be ready again
- check the task manager
- repeat

This doesn't happen in Zed, where changing the value of the settings is automatically picked up and no new processes are spawned.

Is it possible that the extension **doesn't need** a watcher of the configuration settings, and VSCode does everything automatically? 

Eventually, if you close the editor, some zombie processes may remain. 

Still, this doesn't trigger the multiple instances inside the editor, but it could be related.

**Comment by Netail** [comment]
> I might have a suspicion if it could be related to multiple CTRL+S's in a row when fixes are applied? But only a hunch.

Have experienced it with single CTRL + S's as well (With Biome v1 at least)

Related: https://github.com/biomejs/biome-vscode/issues/410

**Comment by nhedger** [comment]
I have found a way to reliably reproduce this somehow.

I'll dig in to see how it can be fixed

**Comment by florianmatz** [comment]
@nhedger How did you manage that, I mean to deterministically reproduce it?

**Comment by nhedger** [comment]
One step at a time. Thanks to @ematipico, I had already suspected the configuration watcher was at fault. This was a tricky one because while looking for this one, I discovered another bug that resulted in multiple _global_ sessions being created sometimes. 

Anyway, PRs incoming

**Comment by nhedger** [comment]
Hopefully, this is now fixed for y'all in v3.1.0 of the extension. If not, let us know.

**Comment by WerdoxDev** [comment]
So unfortunately this is still an issue in v3.1.0 of the extension (v2.0.6 @biome/biomejs)
I still see two biome instances in the formatters list. Saving makes the instances fight, completely messing up the imports.

**Comment by nhedger** [comment]
> So unfortunately this is still an issue in v3.1.0 of the extension (v2.0.6 @biome/biomejs) I still see two biome instances in the formatters list. Saving makes the instances fight, completely messing up the imports.

That's unfortunate.

Are you able to reproduce consistently? If so, can you provide a reproduction repository with instructions? I'll take a look.

**Comment by ematipico** [comment]
No matter what we try, it seems that the issue still happens. Also, we switched the way the client is connected to the server.

I'm beginning to wonder if this is a VSCode issue, because no matter what we do, things don't change

**Comment by nhedger** [comment]
I know for a fact that we solved at least some issue since I no longer see the extra processes, but a way to reproduce would definitely help me debug this

**Comment by Netail** [comment]
> I know for a fact that we solved at least some issue since I no longer see the extra processes, but a way to reproduce would definitely help me debug this

It's really hard to come up with a reproduction. For us it happens really random, maybe once or twice a week. And might be noticed hours later, making it hard to notice why

**Comment by florianmatz** [comment]
I have a file where it happens almost every time, kinda deterministically. I’ll try to provide a reproduction repo, hoping that it will still happen once I strip out all the other stuff… I’ll try to come up with it until monday.

**Comment by florianmatz** [comment]
I have prepared the reproduction repo, you may find it here:

https://github.com/florianmatz/biome-repro

Code is complete nonsense, because I stripped out everything unnecessary. 

However, with that it is possible to replicate the error, where biome messes the code on save. I have also included the vscode settings I use in the real repo. 

### Steps to reproduce 

Same as in the video I've provided in the issue I originally created: [https://www.youtube.com/shorts/4D5-ovverXg](https://www.youtube.com/shorts/4D5-ovverXg)

- Setup the repo (pnpm install)
- Go to file "filterable-table.component.tsx"
- Mark everything from the beginning of the component to the return statement. 

<img width="560" height="908" alt="Image" src="https://github.com/user-attachments/assets/7e7e7640-30a5-4015-b293-ea745b46871a" />

- Hit "Delete"
- Hit "CMD + Save"
-  Checkout the result

<img width="560" height="547" alt="Image" src="https://github.com/user-attachments/assets/b7566dff-cb91-495a-b44a-90fd5b054f24" />


----

Biome 2.0.6
Biome Extension 3.1.0
VSCode 1.101.2

**Comment by florianmatz** [comment]
Any updates / possible to reproduce on your side with the repo, @nhedger ?

**Comment by momargoh** [comment]
Hi @florianmatz, I couldn't reproduce the issue following your instructions and indeed haven't seen it since at least v3.1.0 of the extension, so it appears to be fixed on my machine at least. I tried on my macbook m1 and also inside the node 22 bookworm devcontainer `mcr.microsoft.com/devcontainers/typescript-node:1-22-bookworm`

**Comment by Netail** [comment]
I was able to reproduce it with the latest vscode extension & biome v2.0.6, but no idea how it triggered

**Comment by Jylth** [comment]
I think I found the reason of the multiple biome entries.  So I can reproduce it easily on my side by adding a folder in a monorepo workspace setting which points to the root of the project. So having the following monorepo workspace settings in VSCode:

```
{
  "folders": [
    // this root path is causing the multiple entries of biome to show up
    {
       "name": "Root",
       "path": "."
    },
    {
      "name": "Frontend",
      "path": "./apps/frontend/"
    },
    {
      "name": "Backend",
      "path": "./apps/backend/"
    }
  ]
}
```

Removing the Root path from the folder list fixes it, re-adding it re-shows multiple biome entries. This root path is the only one which causes an issue, any other folder path not pointing to the root of the project does not cause the issue of multiple entries to arise.

**Comment by kerzeld** [comment]
As a workaround dissable the biome extension in the workspace file and enable it in the `Root` folder then it works as expected.

**Comment by erossignon** [comment]
This may help restarting on good ground

```powershell
# kill_biome_all.ps1
# This script repeatedly kills all biome.exe processes until none remain

do {
    $processes = Get-Process -Name "biome" -ErrorAction SilentlyContinue
    if ($processes) {
        Write-Host "Found $($processes.Count) biome.exe process(es). Killing them..."
        $processes | ForEach-Object {
            Write-Host "  Killing process ID: $($_.Id)"
            $_.Kill()
        }
        Start-Sleep -Milliseconds 500 # Wait 500ms for processes to terminate
    }
} while ($processes)
Write-Host "All biome.exe processes have been terminated."````
